src/gpuarray_array_collectives.c

```diff
@@ -0,0 +1,108 @@
+#include "gpuarray/array.h"
+#include "gpuarray/buffer_collectives.h"
+#include "gpuarray/collectives.h"
+#include "gpuarray/error.h"
+
+#include "private.h"
+
+static inline int find_total_elems(const GpuArray* array)
+{
+  unsigned int i;
+  size_t total_elems = 1;
+  for (i = 0; i < array->nd; ++i)
+    total_elems *= array->dimensions[i];
+  return (int) total_elems;
+}
+
+static inline int check_gpuarrays(int times_src, const GpuArray* src,
+                                  int times_dest, const GpuArray* dest, int* count)
+{
+  int count_src, count_dest;
+  count_src = find_total_elems(src);
+  count_dest = find_total_elems(dest);
+  if (times_src * count_src != times_dest * count_dest)
+    return GA_VALUE_ERROR;
+  if (src->typecode != dest->typecode)
+    return GA_VALUE_ERROR;
+  if (!(src->flags & GA_ALIGNED) || !(dest->flags & GA_BEHAVED))
```

Comment:
The GA_BEHAVED check might not do what you want since it will pass if something is aligned OR writeable and I suppose you want aligned AND writable.

The correct code would be `(dest->flags & GA_BEHAVED == GA_BEHAVED)`.  To make sure you get this correctly you can use the `GpuArray_CHKFLAGS()` function